homeassistant/components/hassmpris/__init__.py

```diff
@@ -0,0 +1,104 @@
+"""The MPRIS media playback remote control integration."""
+from __future__ import annotations
+
+import datetime
+
+from cryptography import x509
+from cryptography.hazmat.primitives import serialization
+from cryptography.x509 import Certificate
+import hassmpris_client
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
+
+from .const import (
+    CONF_CLIENT_CERT,
+    CONF_CLIENT_KEY,
+    CONF_HOST,
+    CONF_MPRIS_PORT,
+    CONF_TRUST_CHAIN,
+    DOMAIN,
+    ENTRY_CLIENT,
+    ENTRY_UNLOADERS,
+    LOGGER as _LOGGER,
+)
+
+MIN_TIME_BETWEEN_UPDATES = datetime.timedelta(seconds=5)
+PLATFORMS: list[Platform] = [Platform.MEDIA_PLAYER]
+
+# Requisite for silver quality scale.
+_setup_error_counter: int = 0
+
+
+def _load_cert_chain(chain: bytes) -> list[Certificate]:
+    start_line = b"-----BEGIN CERTIFICATE-----"
+    cert_slots = chain.split(start_line)
+    certificates: list[Certificate] = []
+    for single_pem_cert in cert_slots[1:]:
+        loaded = x509.load_pem_x509_certificate(start_line + single_pem_cert)
+        certificates.append(loaded)
+    return certificates
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up MPRIS media playback remote control from a config entry."""
+    global _setup_error_counter
+
+    client_cert = x509.load_pem_x509_certificate(
+        entry.data[CONF_CLIENT_CERT].encode("ascii"),
+    )
+    client_key = serialization.load_pem_private_key(
+        entry.data[CONF_CLIENT_KEY].encode("ascii"),
+        None,
+    )
+    trust_chain = _load_cert_chain(
+        entry.data[CONF_TRUST_CHAIN].encode("ascii"),
+    )
+
+    clnt = hassmpris_client.AsyncMPRISClient(
+        entry.data[CONF_HOST],
+        entry.data[CONF_MPRIS_PORT],
+        client_cert,
+        client_key,
+        trust_chain,
+    )
+    try:
+        _LOGGER.debug("Pinging the MPRIS agent")
+        await clnt.ping()
+        _LOGGER.info("Successfully pinged the MPRIS agent")
+        _setup_error_counter = 0
+
+    except hassmpris_client.Unauthenticated as exc:
+        raise ConfigEntryAuthFailed(exc) from exc
+    except Exception as exc:
+        if _setup_error_counter == 0:
+            _LOGGER.warning("Cannot ping the MPRIS agent: %s", exc)
+            _setup_error_counter += 1
+        raise ConfigEntryNotReady(str(exc)) from exc
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {
+        ENTRY_CLIENT: clnt,
+        ENTRY_UNLOADERS: [],
+    }
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    return True
+
+
```

Comment:
It looks like a lot of the certificate logic could be moved into the `hassmpris_client` library. We want Home Assistant to know as little as possible about the integration implementation details

Reply:
The thing is, this code (casting certs and keys from bytes to objects) is _only_ needed because HASS insists on storing strings / bytes -- the integration only deals in actual typed Python objects, like `Certificate` and `Key`.  Short of moving these casting functions into helpers within the client, this actually belongs here and not in the MPRIS client.  All of the logic of actually obtaining the certificate and pairing with the remote end is fully sealed up in the MPRIS client.

What should I do?  I'm 100% new to writing HASS code, I need your help.  Thanks in advance.

Reply:
Roger that.  The client module has no provision to store anything on disk -- it is stateless -- so I will factor out the certificate load / save code to their own separate functions and proceed as you advised with regards to mirroring what `homekit` does.